File: scripts/py/filterrules.py
import os
import re
import requests
import traceback

# Define a dictionary that contains the arrays of URLs, begin markers, end markers, and output files
filter_parameters = {
    "source_files": {
        "https://raw.githubusercontent.com/LM-Firefly/Rules/master/PROXY/Amazon.list": "Amazon.list",
        #"https://raw.githubusercontent.com/blackmatrix7/ios_rule_script/master/rule/Surge/Amazon/Amazon.list": "AmazonAll.list",
        #"https://raw.githubusercontent.com/LM-Firefly/Rules/master/PROXY/Google.list": "GoogleAll.list"
    },
    "markers": {
        "## >> Amazon Prime Video": "## >> Audible",
        #"## >> CreateSpace": "## >> End CreateSpace",
        #"## >> Youtube": "## >> End Youtube"
    },
    "lines_to_exclude": {
        "https://raw.githubusercontent.com/blackmatrix7/ios_rule_script/master/rule/Surge/AmazonPrimeVideo/AmazonPrimeVideo.list",
        "example1.example.com",
        "example2.example.com"
    }
}

# Local destination folder path
destination_folder = os.getcwd()

# Download list files and filter from filter_parameters
def fetch_and_filter_content(filter_parameters):
    # Fetch lines to exclude from URLs
    lines_to_exclude_temp = filter_parameters['lines_to_exclude'].copy()
    for line_to_exclude in lines_to_exclude_temp:
        if line_to_exclude.startswith('http://') or line_to_exclude.startswith('https://'):
            try:
                response = requests.get(line_to_exclude, timeout=10)
                if response.status_code == 200:
                    lines = response.text.split('\n')
                    for line in lines:
                        if not line.startswith('#'):
                            filter_parameters['lines_to_exclude'].add(line)
                filter_parameters['lines_to_exclude'].remove(line_to_exclude)
            except requests.exceptions.RequestException as e:
                print(f"An error occurred while fetching the exclude list from '{line_to_exclude}': {e}")
                print(traceback.format_exc())
            except Exception as e:
                print(f"An unexpected error occurred: {e}")
                print(traceback.format_exc())

    # Fetch and filter source files
    for source_url, filtered_file in filter_parameters['source_files'].items():
        try:
            response = requests.get(source_url, timeout=10)

            if response.status_code == 200:
                content = response.text
                for begin, end in filter_parameters['markers'].items():
                    pattern = r'{}.*?(?={})'.format(re.escape(begin), re.escape(end))
                    content = re.sub(pattern, '', content, flags=re.DOTALL)
                
                for line_to_exclude in filter_parameters['lines_to_exclude']:
                    content = content.replace(line_to_exclude, '')

                with open(os.path.join(destination_folder, filtered_file), 'w') as f:
                    f.write(content)
                print(f"File copied from {source_url} filtered and added to {filtered_file}.")
            else:
                print(f"File '{source_url}' not found.")
        except requests.exceptions.RequestException as e:
            print(f"An error occurred while fetching the file from '{source_url}': {e}")
            print(traceback.format_exc())
        except Exception as e:
            print(f"An unexpected error occurred: {e}")
            print(traceback.format_exc())

if __name__ == "__main__":
    fetch_and_filter_content(filter_parameters)

File: scripts/js/filterrules.js
const fs = require('fs').promises
const https = require('https')
const path = require('path')

const filterParameters = {
  sourceFiles: {
    'https://raw.githubusercontent.com/LM-Firefly/Rules/master/PROXY/Amazon.list': 'Amazon.list'
    // "https://raw.githubusercontent.com/blackmatrix7/ios_rule_script/master/rule/Surge/Amazon/Amazon.list": "AmazonAll.list",
    // "https://raw.githubusercontent.com/LM-Firefly/Rules/master/PROXY/Google.list": "GoogleAll.list"
  },
  markers: {
    '## >> Amazon Prime Video': '## >> Audible'
    // "## >> CreateSpace": "## >> End CreateSpace",
    // "## >> Youtube": "## >> End Youtube"
  },
  linesToExclude: [
    'https://raw.githubusercontent.com/blackmatrix7/ios_rule_script/master/rule/Surge/AmazonPrimeVideo/AmazonPrimeVideo.list',
    'example1.example.com',
    'example2.example.com'
  ]
}

const destinationFolder = process.cwd()

function httpsGet (url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = ''

      res.on('data', (chunk) => {
        data += chunk
      })

      res.on('end', () => resolve(data))
      res.on('error', reject)
    })
  })
}

async function fetchAndFilterContent (filterParameters) {
  let linesToExclude = await Promise.all(
    filterParameters.linesToExclude.map(async (line) => {
      if (line.startsWith('http://') || line.startsWith('https://')) {
        try {
          const data = await httpsGet(line)
          return data.split('\n').filter((line) => !line.startsWith('#'))
        } catch (error) {
          console.error(`Failed to fetch data from ${line}: ${error}`)
          return []
        }
      } else {
        return [line]
      }
    })
  )
  linesToExclude = linesToExclude.flat()

  const [sourceUrl, filteredFile] = Object.entries(filterParameters.sourceFiles)[0]
  try {
    let data = await httpsGet(sourceUrl)
    Object.entries(filterParameters.markers).forEach(([begin, end]) => {
      const pattern = new RegExp(`${escapeRegExp(begin)}.*?(?=${escapeRegExp(end)})`, 'gs')
      data = data.replace(pattern, '')
    })

    linesToExclude.forEach((lineToExclude) => {
      data = data.replace(lineToExclude, '')
    })

    await fs.writeFile(path.join(destinationFolder, filteredFile), data)
    console.log(`File copied from ${sourceUrl} filtered and added to ${filteredFile}.`)
  } catch (error) {
    console.error(`Failed to process file ${sourceUrl}: ${error}`)
  }
}

function escapeRegExp (string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

fetchAndFilterContent(filterParameters)
